### PR TITLE
Merge/foundation release/1.11

### PR DESCRIPTION
**Pull request by meowsbits**
Merges ethereum/go-ethereum branch `release/1.11`  to core-geth `master`.

Significant changes include
- Ropsten is dead and gone; Sepolia replaces it as the canonical testnet.
- [Shangahai feature support](https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/shanghai.md) is added;
  - Introducing time-based fork definitions
- `eth/68` is the new best-default protocol
- 2 `debug_` API methods gone, 4 new ones introduced.
- Pebble DB.
- `personal` API is hashly deprecated, requiring opt-back-in with `--rpc.enabledeprecatedpersonal`.
- Removed iOS and Android builds

Things I'd like to consider for this or next release
- [x] Remove all Parity, Aleth, and Multigeth configuration support. I don't think many people use this. Being abandoned projects, their chain spec schemas have long since not supported configurations for features post-Byzantium (so they could not be well-supported).

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -4,7 +4,7 @@ ARG VERSION=""
 ARG BUILDNUM=""
 
 # Build Geth in a stock Go builder container
-FROM golang:1.18-alpine as builder
```
Comment (ziogaschr) [inline Dockerfile:7]: Not important, just a heads up. `go.mod` suggests go1.19, which I used for testing. Dockerfile's are using go1.20, which is probably fine, but good to give them a test as well as they might create a regression.

File: cmd/geth/dbcmd.go [hunk 2]
```diff
@@ -704,129 +694,22 @@ func showMetaData(ctx *cli.Context) error {
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "Error accessing ancients: %v", err)
 	}
-	pp := func(val *uint64) string {
```
Comment (ziogaschr) [inline cmd/geth/dbcmd.go:707]: Do you know the reason behind the changes in this file?
Will research upstream tomorrow, but was thinking if it is with regards **pebble**, or transition to PoS

File: consensus/ethash/ethash.go [hunk 3]
```diff
@@ -275,55 +291,11 @@ type cache struct {
...
-func newCache(epoch uint64, epochLength uint64) interface{} {
+// newCache creates a new ethash verification cache.
+func newCache(epoch uint64, epochLength uint64) *cache {
 	return &cache{epoch: epoch, epochLength: epochLength}
 }
 
-// isBadCache checks a given caches/datsets keccak256 hash against bad caches (ecip-1099)
-// this is incase the client has already written non-ecip1099 caches to disk,
-// instead of blindly trusting as seedhashes/filename match, compare checksums.
-func isBadCache(epoch uint64, epochLength uint64, data []uint32) (bool, string) {
```
Comment (ziogaschr) [inline consensus/ethash/ethash.go:287]: Good to have this logic out.

File: core/blockchain.go [hunk 4]
```diff
@@ -2310,72 +2389,58 @@ func (bc *BlockChain) skipBlock(err error, it *insertIterator) bool {
...
 // User can use flag `txlookuplimit` to specify a "recentness" block, below
 // which ancient tx indices get deleted. If `txlookuplimit` is 0, it means
 // all tx indices will be reserved.
 //
-// The user can adjust the txlookuplimit value for each launch after fast
-// sync, Geth will automatically construct the missing indices and delete
-// the extra indices.
-func (bc *BlockChain) maintainTxIndex(ancients uint64) {
+// The user can adjust the txlookuplimit value for each launch after sync,
+// Geth will automatically construct the missing indices or delete the extra
```
Comment (ziogaschr) [inline core/blockchain.go:2438]: I didn't knew that changing the txLookupLimit can recreate the indices. Though at some point @meowsbits referred to this possibility.

File: core/forkid/forkid.go [hunk 5]
```diff
@@ -42,6 +42,12 @@ var (
 	ErrLocalIncompatibleOrStale = errors.New("local incompatible or needs update")
 )
 
+// timestampThreshold is the Ethereum mainnet genesis timestamp. It is used to
+// differentiate if a forkid.next field is a block number or a timestamp. Whilst
+// very hacky, something's needed to split the validation during the transition
+// period (block forks -> time forks).
+const timestampThreshold = 1438269973
```
Comment (ziogaschr) [inline core/forkid/forkid.go:49]: This is safe for ETC, but let's keep an eye on it

File: core/state/metrics.go [hunk 6]
```diff
@@ -19,10 +19,12 @@ package state
 import "github.com/ethereum/go-ethereum/metrics"
 
 var (
-	accountUpdatedMeter        = metrics.NewRegisteredMeter("state/update/account", nil)
-	storageUpdatedMeter        = metrics.NewRegisteredMeter("state/update/storage", nil)
-	accountDeletedMeter        = metrics.NewRegisteredMeter("state/delete/account", nil)
-	storageDeletedMeter        = metrics.NewRegisteredMeter("state/delete/storage", nil)
-	accountTrieCommittedMeter  = metrics.NewRegisteredMeter("state/commit/accountnodes", nil)
```
Comment (ziogaschr) [inline core/state/metrics.go:26]: Are we using those at our metrics dashboard?

File: core/txpool/txpool.go [hunk 7]
```diff
@@ -358,14 +370,8 @@ func (pool *TxPool) loop() {
 		// Handle ChainHeadEvent
 		case ev := <-pool.chainHeadCh:
 			if ev.Block != nil {
-				if pool.chainconfig.IsEnabled(pool.chainconfig.GetEIP2Transition, ev.Block.Number()) {
```
Comment (ziogaschr) [inline core/txpool/txpool.go:361]: Do we want to keep those checks in core-geth, as we support more chains that might want to have those eips off (we don't support such chains, but still)

File: core/vm/jump_table.go [hunk 8]
```diff
@@ -201,12 +201,18 @@ func instructionSetForConfig(config ctypes.ChainConfigurator, isPostMerge bool,
 			maxStack:    maxStack(0, 1),
 		}
 	}
+	if config.IsEnabledByTime(config.GetEIP3855TransitionTime, bt) {
```
Comment (ziogaschr) [inline core/vm/jump_table.go:204]: For later, if we activate future EIPs in ETC, we have to add logic for checking with `isEnabled` as we don't use the time logic

File: core/vm/runtime/runtime.go [hunk 9]
```diff
@@ -118,11 +114,20 @@ func Execute(code, input []byte, cfg *Config) ([]byte, *state.StateDB, error) {
...
-	}
 
+		// Berlin
+		// https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/berlin.md
+		// EIP-2930: Optional access lists
+		eip2930f = cfg.ChainConfig.IsEnabled(cfg.ChainConfig.GetEIP2930Transition, vmenv.Context.BlockNumber)
+
+		// Shanghai
+		// EIP-3651: Warm coinbase
+		eip3651f = cfg.ChainConfig.IsEnabledByTime(cfg.ChainConfig.GetEIP3651TransitionTime, &vmenv.Context.Time)
```
Comment (ziogaschr) [inline core/vm/runtime/runtime.go:125]: Check with `isEnabled` as well

File: core/vm/runtime/runtime.go [hunk 10]
```diff
@@ -151,10 +155,20 @@ func Create(input []byte, cfg *Config) ([]byte, common.Address, uint64, error) {
...
 		sender = vm.AccountRef(cfg.Origin)
+
+		// Berlin
+		// https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/berlin.md
+		// EIP-2930: Optional access lists
+		eip2930f = cfg.ChainConfig.IsEnabled(cfg.ChainConfig.GetEIP2930Transition, vmenv.Context.BlockNumber)
+
+		// Shanghai
+		// EIP-3651: Warm coinbase
+		eip3651f = cfg.ChainConfig.IsEnabledByTime(cfg.ChainConfig.GetEIP3651TransitionTime, &vmenv.Context.Time)
```
Comment (ziogaschr) [inline core/vm/runtime/runtime.go:166]: Check with `isEnabled` as well

**Review by ziogaschr — COMMENTED** [review]
This review is not finished. I am going to submit it though, and continue with PR comments from now. I reviewed up to `eth/downloader/beaconsync.go`.

**Comment by ziogaschr** [comment]
We will have to search globally for`isEnabledByTime` and modify accordingly.

File: eth/peer.go [hunk 11]
```diff
@@ -24,6 +24,17 @@ import (
 	"github.com/ethereum/go-ethereum/eth/protocols/snap"
 )
 
+/* PTAL(meowsbits)
+https://github.com/ethereum/go-ethereum/pull/26804
```
Comment (ziogaschr) [inline eth/peer.go:28]: I vote on reverting this PR @meowsbits

File: eth/protocols/eth/protocol.go [hunk 12]
```diff
@@ -40,11 +41,11 @@ const ProtocolName = "eth"
 
 // ProtocolVersions are the supported versions of the `eth` protocol (first
 // is primary).
-var ProtocolVersions = []uint{ETH67, ETH66}
+var ProtocolVersions = []uint{ETH68, ETH67, ETH66}
```
Comment (ziogaschr) [inline eth/protocols/eth/protocol.go:44]: NOTE: Also check **SupportedProtocolVersions**

File: eth/protocols/snap/handler.go [hunk 13]
```diff
@@ -487,19 +489,13 @@ func ServiceGetTrieNodesQuery(chain *core.BlockChain, req *GetTrieNodesPacket, s
...
 
-	accTrie, err := trie.NewStateTrie(common.Hash{}, req.Root, triedb)
+	accTrie, err := trie.NewStateTrie(trie.StateTrieID(req.Root), triedb)
 	if err != nil {
 		// We don't have the requested state available, bail out
 		return nil, nil
 	}
+	// The 'snap' might be nil, in which case we cannot serve storage slots.
 	snap := chain.Snapshots().Snapshot(req.Root)
-	if snap == nil {
```
Comment (ziogaschr) [inline eth/protocols/snap/handler.go:496]: @meowsbits let's discuss in call if we want to keep this in

File: eth/tracers/js/goja.go [hunk 14]
```diff
@@ -238,7 +250,7 @@ func (t *jsTracer) CaptureStart(env *vm.EVM, from common.Address, to common.Addr
 	t.ctx["block"] = t.vm.ToValue(env.Context.BlockNumber.Uint64())
 	// Update list of precompiles based on current block
 	t.activePrecompiles = env.ActivePrecompiles()
-	t.ctx["intrinsicGas"] = t.vm.ToValue(t.gasLimit - gas)
+	t.ctx["intrinsicGas"] = t.vm.ToValue(t.gasLimit - gas) // PTAL-meowsbits Is this supposed to be here?
```
Comment (ziogaschr) [inline eth/tracers/js/goja.go]: Yes at the moment as tracers need this to be available.

```suggestion
	t.ctx["intrinsicGas"] = t.vm.ToValue(t.gasLimit - gas)
```

File: eth/tracers/js/goja.go [hunk 15]
```diff
@@ -255,10 +267,11 @@ func (t *jsTracer) CaptureState(pc uint64, op vm.OpCode, gas, cost uint64, scope
...
 	log.stack.stack = scope.Stack
 	log.contract.contract = scope.Contract
-	log.pc = uint(pc)
-	log.gas = uint(gas)
-	log.cost = uint(cost)
-	log.depth = uint(depth)
+	log.pc = pc
+	log.gas = gas
+	log.cost = cost
+	log.refund = t.env.StateDB.GetRefund()
```
Comment (ziogaschr) [inline eth/tracers/js/goja.go:272]: @ziogaschr this is probably not needed, as the Parity JS tracer has been moved out to a native tracer

File: params/confp/convert.go [hunk 16]
```diff
@@ -24,11 +24,29 @@ import (
 	"github.com/ethereum/go-ethereum/params/types/ctypes"
 )
 
-// Automagically translate between [Must|]Setters and Getters.
-func Convert(from, to interface{}) error {
-	// Interfaces must be either ChainConfigurator or GenesisBlocker.
+// CloneChainConfigurator creates a copy of the given ChainConfigurator.
+func CloneChainConfigurator(from ctypes.ChainConfigurator) (ctypes.ChainConfigurator, error) {
```
Comment (ziogaschr) [inline params/confp/convert.go:28]: Nice work in this file

File: params/vars/protocol_params.go [hunk 17]
```diff
@@ -41,7 +41,7 @@ var (
 var (
 	// SupportedProtocolVersions are the supported versions of the `eth` protocol (first
 	// is primary).
-	SupportedProtocolVersions = []uint{67, 66}
+	SupportedProtocolVersions = []uint{68, 67, 66}
```
Comment (ziogaschr) [inline params/vars/protocol_params.go:44]: Greatm **68** has been added here as well

File: core/vm/jump_table.go (hunk 8 above)
Comment (meowsbits) [reply core/vm/jump_table.go:204]: Not sure what you mean here... 
- ETC does not have any EIPs that require time-based activation (but ETH does and this patch supports that). 
- the configurator interface defines `IsEnabledByTime` adjacent to `IsEnabled` (by block); 
  - configurator convention expects time-based activation configuration values (fields) to be named like `EIP424242FTime` (suffix=`Time`), vs.  `EIP2FBlock` (suffix=`Block`). 

What logic needs to be yet added?

File: eth/peer.go (hunk 11 above)
Comment (meowsbits) [reply eth/peer.go:28]: SGTM

File: core/vm/jump_table.go (hunk 8 above)
Comment (ziogaschr) [reply core/vm/jump_table.go:204]: Sorry for not explaining better. I finished the first review of full PR an hour ago and I am going to check back comments and add commits for parts needed.

Based on this, my idea is that core-geth have to support both `ByBlock` and `ByTime` logic everywhere. Upstream is clear that will use `ByTime` from now on.

For future EIPs that ETC or other chains using core-geth will utilise on their network they might want to enable an EIP based on the current `ByBlock` logic.

My suggestion is to have something like: `if config.IsEnabledByTime(config.GetEIP3855TransitionTime, bt) || config.IsEnabled(config.GetEIP3855TransitionNumber, bn) {`

File: core/txpool/txpool.go (hunk 7 above)
Comment (ziogaschr) [reply core/txpool/txpool.go:361]: Follow the comments here as it's the same case https://github.com/etclabscore/core-geth/pull/531#discussion_r1157160831

File: eth/protocols/eth/protocol.go (hunk 12 above)
Comment (ziogaschr) [reply eth/protocols/eth/protocol.go:44]: This is indeed being handled in this PR [here](https://github.com/etclabscore/core-geth/pull/531#discussion_r1161662777)

File: core/vm/runtime/runtime.go (hunk 10 above)
Comment (ziogaschr) [reply core/vm/runtime/runtime.go:166]: Follow comments [here](https://github.com/etclabscore/core-geth/pull/531#discussion_r1157160831) (same logic)

File: core/vm/runtime/runtime.go (hunk 9 above)
Comment (ziogaschr) [reply core/vm/runtime/runtime.go:125]: Follow comments [here](https://github.com/etclabscore/core-geth/pull/531#discussion_r1157160831) (same logic)

File: eth/tracers/js/goja.go (hunk 14 above)
Comment (ziogaschr) [reply eth/tracers/js/goja.go]: Indeed this can go now. But some stuff have to be changed on some tracers as well. I will make a commit to fix this one

Comment (ziogaschr) [reply eth/tracers/js/goja.go]: Resolved by https://github.com/etclabscore/core-geth/pull/531/commits/a253f932f68cc444c36dffe2ceb2f1e5c3935571

File: eth/tracers/js/goja.go (hunk 15 above)
Comment (ziogaschr) [reply eth/tracers/js/goja.go:272]: We still need this one

File: cmd/geth/dbcmd.go (hunk 2 above)
Comment (ziogaschr) [reply cmd/geth/dbcmd.go:707]: They have been extracted to another file. Check [here](https://github.com/ethereum/go-ethereum/commit/f7661a662a6e0743b440956e4e15bdae1ba3821e)

File: core/state/metrics.go (hunk 6 above)
Comment (meowsbits) [reply core/state/metrics.go:26]: Indeed, yes. Nice catch! So it looks like `state/commit/accountnodes` becomes `state/update/accountnodes`, and then 3 other gauges are added. That'll be a TODO over at the dashboards where applicable, but not a major change..

**Pull request merged**